instrumentation/apache-httpclient/apache-httpclient-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v4_0/ApacheHttpClientRequest.java

```diff
@@ -121,4 +121,29 @@ public Integer getPeerPort() {
         return null;
     }
   }
+
+  @Nullable
+  private static URI getCalculatedUri(HttpHost httpHost, HttpRequest httpRequest) {
+    final URI uri;
+    try {
+      // this can be relative or absolute
+      uri = new URI(httpRequest.getRequestLine().getUri());
```

Comment:
What do you think about recognizing a scheme (`contains("://")`) to avoid parsing (and losing fidelity?) an absolute URI.

By the way I had noticed randomly the issue about concatenating two absolute URIs together and the instanceof check I added fixed that, is there another case where we can have two absolute URIs? I guess we can document it in #3697 where I guess we've found ourselves eating some Apache-flavored spaghetti

Reply:
By the way here's what Brave does, which of course may also be missing some corner cases, but for reference it seems a bit different than either of our code branches actually

https://github.com/openzipkin/brave/blob/master/instrumentation/httpclient/src/main/java/brave/httpclient/TracingProtocolExec.java#L102

Reply:
spaghetti overload 😋

> What do you think about recognizing a scheme (`contains("://")`) to avoid parsing (and losing fidelity?) an absolute URI.

Yeah, I was just noticing our (unnecessary) URL constructions while profiling and thinking to add some type of "url builder" to our instrumentation api to optimize. Will add tracking issue.